### PR TITLE
Universal MIDI Packet Format (MIDI2.0)

### PR DESCRIPTION
**Pull request by psmokotnin**
MIDI Event changed to UMP format from MIDI2.0 specification.
All events incoming into the app converted to MIDI2.0
All MIDI2.0 event converted to MIDI1.0 when goes out to the ports.
VST takes from MIDI2.0 basic attribute for NoteOn/NoteOff events

### CODE REVIEW

File: framework/midi/miditypes.h [hunk 1]
```diff
@@ -27,18 +27,15 @@
 #include <map>
 #include <functional>
 #include <set>
-#include <array>
 #include <cassert>
 #include "async/channel.h"
-
-#include "audio/midi/event.h"
+#include "midi/midievent.h"
```
Comment (igorkorsukov) [inline framework/midi/miditypes.h:32]: better just `#include "midievent.h"`

File: framework/vst/internal/eventlist.cpp [hunk 2]
```diff
@@ -51,22 +51,36 @@ tresult EventList::getEvent(int32 index, Event& e)
...
         break;
 
-    case midi::EventType::ME_NOTEOFF:
+    case midi::Event::Opcode::NoteOff:
         e.type = Event::kNoteOffEvent;
-        e.noteOff.pitch = midiEvent.note();
+        e.noteOff.noteId = midiEvent.note();
+        e.noteOff.channel = midiEvent.channel();
+        e.noteOff.pitch = midiEvent.pitchNote();
+        e.noteOff.tuning = midiEvent.pitchTuning() * 100;
```
Comment (igorkorsukov) [inline framework/vst/internal/eventlist.cpp:78]: Why multiply by 100?
You need to add a comment or, better, add a constant with a meaningful name

File: framework/midi/midievent.h [hunk 3]
```diff
@@ -0,0 +1,956 @@
...
+            return p.uint32;
+        }
+        return 0;
+    }
+
+    bool operator ==(const Event& other) const { return m_data == other.m_data; }
+    bool operator !=(const Event& other) const { return !operator==(other); }
+    operator bool() const {
+        return operator!=(NOOP());
+    }
```
Comment (igorkorsukov) [inline framework/midi/midievent.h:140]: Is this operator really needed?
Would expect to see a method that `isValid()` (only more complex)

File: framework/midi/midievent.h [hunk 4]
```diff
@@ -0,0 +1,956 @@
...
+
+    bool isChannelVoice() const { return messageType() == ChannelVoice10 || messageType() == ChannelVoice20; }
+    bool isChannelVoice20() const { return messageType() == ChannelVoice20; }
+    bool isMessageTypeIn(std::set<MessageType> types) const { return types.find(messageType()) != types.end(); }
```
Comment (igorkorsukov) [inline framework/midi/midievent.h:144]: better pass `const std::set<MessageType>& types`

File: framework/midi/midievent.h [hunk 5]
```diff
@@ -0,0 +1,956 @@
...
+    bool isOpcodeIn(std::set<Opcode> opcodes) const { return opcodes.find(opcode()) != opcodes.end(); }
```
Comment (igorkorsukov) [inline framework/midi/midievent.h:145]: better pass `const std::set<Opcode>& opcodes`

File: framework/midi/midievent.h [hunk 6]
```diff
@@ -0,0 +1,956 @@
...
+    void setPerNoteReset(bool value)
+    {
+        assertOpcode({ PerNoteManagement });
+        assertMessageType({ ChannelVoice20 });
+        m_data[0] &= 0xFFFFFFFE;
+        m_data[0] |= value;
+    }
+
+    //!convert ChannelVoice from MIDI2.0 to MIDI1.0
+    std::list<Event> toMIDI1_0() const
```
Comment (igorkorsukov) [inline framework/midi/midievent.h:589]: fromMIDI**10**Package, but toMIDI**1_0** 

I think it's better to do it uniformly

**Comment by igorkorsukov** [comment]
Very cool!

File: framework/midi/midievent.h (hunk 3 above)
Comment (psmokotnin) [reply framework/midi/midievent.h:140]: NOOP - is a valid (and standard) MIDI message. In the same time, it is a NULL message, that may not be processed and can be skipped.
example of usage:
```
if (e) {
    m_eventReceived.send({ timing, e });
}
```
It won't litter the event stack.

File: framework/vst/internal/eventlist.cpp (hunk 2 above)
Comment (psmokotnin) [reply framework/vst/internal/eventlist.cpp:78]: It is a conversion from semitones to cents. 
Fixed. Added method:
```
//! return tuning in cents
float pitchTuningCents() const
{
    return pitchTuning() * 100;
}
```

File: framework/midi/midievent.h (hunk 6 above)
Comment (psmokotnin) [reply framework/midi/midievent.h:589]: Fixed

File: framework/midi/miditypes.h (hunk 1 above)
Comment (psmokotnin) [reply framework/midi/miditypes.h:32]: fixed

File: framework/midi/midievent.h (hunk 3 above)
Comment (igorkorsukov) [reply framework/midi/midievent.h:140]: This is where I see problems

```
Event e1;
if (e1) {
  // not pass
}

Event e2;
e2.setMessageType(...);
if (e2) {
  // pass
  // but e2 not valid
}
```

Comment (igorkorsukov) [reply framework/midi/midievent.h:140]: i.e. it is not obvious what the true means for the event

File: framework/midi/midievent.h (hunk 4 above)
Comment (psmokotnin) [reply framework/midi/midievent.h]: fixed

File: framework/midi/midievent.h (hunk 5 above)
Comment (psmokotnin) [reply framework/midi/midievent.h]: fixed

File: framework/midi/midievent.h (hunk 3 above)
Comment (psmokotnin) [reply framework/midi/midievent.h:140]: Yes, I missed that case. I'll add a check for message's validation.

Comment (psmokotnin) [reply framework/midi/midievent.h:146]: fixed

**Review by igorkorsukov — APPROVED** [review]

**Pull request merged**